Spin up CloudWatch for CloudTrail logs

ahorner opened the pull request:
As a first step toward improving our AWS security defaults, we'd like
to enable CloudWatch any time we're spinning up a CloudTrail trail.
This will give us a lot of freedom moving forward to set up alerts,
monitoring, et cetera.

As part of this change, we're also making a few tweaks to our defaults
for the `cloudtrail` module:

- Enable logging for the CloudTrail log bucket
- Enable log file validation for CloudTrail logs
- Use an `aws_iam_policy_document` data source for our bucket policy

There are a few best practices we're _not_ going to try to implement
right now -- encrypting our log bucket with a KMS key, for instance.
But this should be a good step in the right direction to make security
a first-class citizen of our infrastructure management.

It's _entirely_ possible (even likely) that there are better ways to 
accomplish some of this, or a more appropriate way to organize these 
resources. This PR is primarily intended to start the conversation.

ahorner committed:
Spin up CloudWatch for CloudTrail logs

As a first step toward improving our AWS security defaults, we'd like
to enable CloudWatch any time we're spinning up a CloudTrail trail.
This will give us a lot of freedom moving forward to set up alerts,
monitoring, et cetera.

As part of this change, we're also making a few tweaks to our defaults
for the `cloudtrail` module:

- Enable logging for the CloudTrail log bucket
- Enable log file validation for CloudTrail logs
- Use an `aws_iam_policy_document` data source for our bucket policy

There are a few best practices we're _not_ going to try to implement
right now -- encrypting our log bucket with a KMS key, for instance.
But this should be a good step in the right direction to make security
a first-class citizen of our infrastructure management.

phoolish commented:
@ahorner do we have any clue about costs? If we don't or it is cost prohibitive for some of our smaller clients, we might want to make this optional.

ahorner commented:
The [free tier](https://aws.amazon.com/cloudwatch/pricing/#Free_tier) for CloudWatch is pretty generous. I think making these resources optional is a fine idea, but when I was working through this, the optional resources were getting pretty ugly. 

It probably makes more sense to extract `cloudwatch` into its own module, and just expose the group/role ARNs as variables for `cloudtrail`. I'll take a pass at that.

ahorner committed:
Pull CloudWatch into its own terraform module

While there are some concerns that we don't necessarily want to add
CloudWatch for every account -- and I stand by the initial concept
there -- there's a much larger concern about the composability of the
current approach, which doesn't really allow for resources to actually
be added on top of CloudWatch for purposes of alerting, etc.

To address this larger concern, we're extracting the creation of
CloudWatch resources into a separate module, which gives us the Role
and Group ARNs (the only two meaningful pieces of information for
CloudTrail) as outputs, and allows for customization of the list of
services that CloudWatch is allowed to collect information about.

vandrijevik commented:
Hey @ahorner, any thoughts or activity here? I’m seeing a non-empty Terraform plan for one of the clients where I think this would help, so just wondering if we could resolve that by merging this PR, if you’re done with the refactoring you wanted to do…

ahorner commented:
My last commit included the addition of the new CloudWatch module and necessary CloudTrail variables. If everything looks copacetic, we should be able to merge this and start incorporating it into our existing terraform plans, where appropriate.

Apologies if you were waiting on a follow-up comment; I assumed Github would inform everybody watching the PR that I'd pushed a new commit.

vandrijevik commented:
@ahorner I think it may have gotten lost in the mix for me, and/or GitHub is a bit confused on this PR since it keeps showing me a badge _New changes since you last viewed_ with all the old changes… In any case, 👍 and thanks!
